okhttp-testing-support/src/main/java/okhttp3/testing/PlatformRule.kt

```diff
@@ -50,12 +86,54 @@ open class PlatformRule @JvmOverloads constructor(
     }
   }
 
-  override fun after() {
+  fun resetPlatform() {
     if (platform != null) {
       Platform.resetForTests()
     }
   }
 
+  @JvmOverloads
+  fun expectFailure(
+    versionMatcher: Matcher<out Any>,
+    failureMatcher: Matcher<out Any> = CoreMatchers.anything()
+  ) {
+    versionChecks.add(Pair(versionMatcher, failureMatcher))
+  }
+
+  fun expectFailure(
+    versionMatcher: Matcher<out Any>,
+    expectedType: Class<*>
+  ) {
+    versionChecks.add(Pair(versionMatcher, isA(expectedType)))
+  }
+
+  fun rethrowIfNotExpected(e: Throwable) {
+    versionChecks.forEach { (versionMatcher, failureMatcher) ->
+      if (versionMatcher.matches(VersionInfo)) {
+        if (!failureMatcher.matches(e)) {
+          throw e
+        }
+
+        return
+      }
+    }
+
+    throw e
+  }
+
+  fun failIfExpected() {
+    versionChecks.forEach { (versionMatcher, failureMatcher) ->
+      if (versionMatcher.matches(VersionInfo)) {
+        val description = StringDescription()
```

Comment:
For follow up I’d love to find something other than Hamcrest. It’s so awkward.

Reply:
Could be simple kotlin functions if we move some (non API enforcing) tests to kotlin.